src/main/kotlin/no/skatteetaten/aurora/boober/service/resourceprovisioning/S3StorageGridProvisioner.kt

```diff
@@ -0,0 +1,195 @@
+        }
+
+        val errors = requests.filter { !it.result.success }
+
+        val failedRequests = errors.filter { it.result.reason.endState }.takeIf { it.isNotEmpty() }
+            ?.also { logger.debug("ApplicationDeployment ${adc.name} failed ${it.size} SGOA request(s)") }
+        failedRequests?.let { throw S3ProvisioningException("${it.size} StorageGridObjectArea request(s) failed. Check status with \"oc get sgoa -o yaml\"") }
+
+        val timedOutRequests = errors.filter { !it.result.reason.endState }.takeIf { it.isNotEmpty() }
+            ?.also { logger.debug("ApplicationDeployment ${adc.name} timed out ${it.size} SGOA request(s)") }
+        timedOutRequests?.let { throw S3ProvisioningException("Timed out waiting for ${it.size} StorageGridObjectArea request(s) to complete. Check status with \"oc get sgoa -o yaml\"") }
+    }
+
+    private data class SgoaAndResult(
+        val sgoa: StorageGridObjectArea,
+        var result: StorageGridObjectAreaStatusResult = StorageGridObjectAreaStatusResult()
+    )
+
+    private fun openShiftGetObject(resource: HasMetadata): OpenShiftResponse {
+        val namespace = resource.metadata.namespace
+        val cmd = openShiftCommandService.createOpenShiftCommand(namespace, resource)
+            .copy(operationType = OperationType.GET)
+        return openShiftClient.performOpenShiftCommand(namespace, cmd)
+    }
+
+    private fun HerkimerService.getObjectAreaAdminCredentials(adc: AuroraDeploymentSpec): List<ObjectAreaWithCredentials> {
+
+        val sgoaResources = getObjectAreaResourcesIndex(adc)
+        return adc.s3ObjectAreas.map { objectArea ->
+
+            val objectAreaResourceName = objectArea.specifiedAreaKey
+            val sgoaResource = sgoaResources[objectAreaResourceName]
+                ?: throw S3ProvisioningException("Unable to find resource of kind ${ResourceKind.StorageGridObjectArea} with name $objectAreaResourceName for AuroraDeploymentSpec ${adc.name}")
+
+            ObjectAreaWithCredentials(objectArea, sgoaResource.adminClaimCredentials)
+        }
+    }
+
+    private fun HerkimerService.getObjectAreaResourcesIndex(adc: AuroraDeploymentSpec): Map<String, ResourceHerkimer> =
+        getClaimedResources(adc.applicationDeploymentId, ResourceKind.StorageGridObjectArea)
+            .associateBy { it.name }
+
+    private val ResourceHerkimer.adminClaimCredentials
+        get(): StorageGridCredentials {
+            val claimName = "ADMIN"
+            val adminClaim = claims.find { it.name == claimName }
+                ?: throw S3ProvisioningException("Unable to find claim with name $claimName for resource of kind ${this.kind} with name ${this.name}")
+            return adminClaim.credentials.convert<StorageGridCredentials>()
+                .copy(bucketRegion = defaultBucketRegion)
+        }
+
+    private fun createSgoaResource(adc: AuroraDeploymentSpec, objectArea: S3ObjectArea): StorageGridObjectArea {
```

Comment:
Her mangler ownerReference og endel "common labels". Er det ok, inntil videre? Hva vil en god måte være å få "tak" i denne infoen? StorageGridObjectArea inngår ikke i den vanlige "flyten" med features, etc.

Reply:
Egentlig burde vel kanskje ApplicationDeployment bli skrevet først? Ser at det kan kreve litt innsats, men på sikt bør vi kanskje ha en felles måte å definere at noen features skal kjøres i steg 1, mens andre skal kjøre i steg 2 (vanlig generate og modify i dag.) ved hvert steg så vil disse commonFeatures kjøres for å sette diverse labels og slikt. Men jeg har ikke komplett oversikt over features så det er sikkert noen som ikke passer inn i denne modellen. Har du sett for deg noe når du har holdt på? 
"steg" er kun brukt for å illustrere

Reply:
Kunne vi skrevet SGOA to ganger og andre gang hadde den fått den infoen den trenger. Da må generate returnere SGOA CR i tillegg vel?

Reply:
Hvis det hadde gått ville det vært en low effort greie vil jeg tro.

Reply:
Ikke dumt. Det er absolutt verdt en test 👍